### PR TITLE
KYLIN-4104 Support multi jdbc pushdown runners to execute query/update

### PR DESCRIPTION
**Pull request by weibin0516**

### CODE REVIEW

File: core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java [hunk 1]
```diff
@@ -1745,6 +1746,17 @@ public String getPushDownRunnerClassName() {
         return getOptional("kylin.query.pushdown.runner-class-name", "");
     }
 
+    public List<String> getPushDownRunnerIds() {
+        List<String> ids = Lists.newArrayList();
+        String idsStr = getOptional("kylin.query.pushdown.runner.ids", "");
```
Comment (nichunen) [inline core-common/src/main/java/org/apache/kylin/common/KylinConfigBase.java]: should be "kylin.query.pushdown.runner-ids"

File: query/src/main/java/org/apache/kylin/query/util/PushDownExecutor.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            SQLException sqlException, boolean isSelect, boolean isPrepare) throws Exception {
+        Pair<List<List<String>>, List<SelectedColumnMeta>> ret = null;
+
+        for (int i = 0; i < ids.size(); i++) {
+            String id = ids.get(i);
+            PushDownRunnerJdbcImpl runner = new PushDownRunnerJdbcImpl();
+            runner.initById(kylinConfig, id);
+
+            try {
+                ret = queryBySingleRunner(runner, project, sql, defaultSchema, sqlException,
```
Comment (nichunen) [inline query/src/main/java/org/apache/kylin/query/util/PushDownExecutor.java]: Why not return the result from the first engine immediately if more than one engines are available?

Comment (weibin0516) [reply query/src/main/java/org/apache/kylin/query/util/PushDownExecutor.java]: @nichunen , thanks to review, it has been fixed

**Comment by nichunen** [comment]
Hi @weibin0516 

The pr approved. Would you please submit another pr for the update of the document, it's on the document branch, in "website/_docs31"

**Comment by weibin0516** [comment]
> Hi @weibin0516
> 
> The pr approved. Would you please submit another pr for the update of the document, it's on the document branch, in "website/_docs31"

Hi, @nichunen , document pr is https://github.com/apache/kylin/pull/1027

**Comment by zhangayqian** [comment]
### Check list
- [ ] Step 1. [**OPTIONAL**]Understand background/root cause/design basically(one hour EST) . Its issue type is 
    - [ ] bug fix
    - [x] new feature
    - [ ] enhancement
- [ ] Step 2. Test cases are designed and documented (30 minutes EST).
- [ ] Step 3. Prepare specific env, for example:
    - mock data (maybe some specific data type)
    - test env (maybe install a RDBMS instance).
- [ ] Step 4. Verify and make sure test cases passed.
- [ ] Step 5. Paste manual important steps and screenshots here (20 minutes EST). 
    - If you find difficulty in pick most important evidence, please attach diagnosis package. 
- [ ] Step 6. Do more check in user perspective (20 minutes EST)
    - [ ] Doc need be updated? And if it is updated? Ask help for release manager if so.
    - [ ] If it is a breaking change so we should notify Kylin community? Ask help for release manager if so.
- [ ] Step 7.  Summarize this test (20 minutes EST).

### Total estimate
- 4 hours for small issue from optimistic estimate. 

### Note
If you find some unexpected and **unrelated** error/mistake, please DO report it if it is truly a mistake, I think we may research and fix it in the future.

If you find background information/root cause analysis is not complete or ambiguous, please try to contact to author or do a quick research and record what you find. It is a good chance to learn something interesting.

If you find it is not easy to design testcase, please notify release manager.

**Comment by zhangayqian** [comment]
## Design test cases

### Prepare test env
- Prepare Presto env and import kylin_sales
- bin/kylin.sh start
### Test case
#### Case-1
- Ensure hive query engine pushdown is available
  - Config kylin query pushdown to hive according to https://github.com/apache/kylin/pull/1027.
  - Restart kylin and disable all the cube in `learn_kylin`
  - execute `select PART_DT, sum(price) from kylin_sales where PART_DT<'2012-01-15'` in project `learn_kylin`.
#### Expected result
- SQL is pushed down to hive and returns the correct result.
#### Case-2
- Ensure presto query engine pushdown is available
  - Config kylin query pushdown to presto according to https://github.com/apache/kylin/pull/1027.
  - Restart kylin and disable all the cube in `learn_kylin`
  - execute `select PART_DT, sum(price) from kylin_sales where PART_DT<'2012-01-15'` in project `learn_kylin`.
#### Expected result
- SQL is pushed down to presto and returns the correct result.
#### Case-3
- Ensure multi Jdbc pushdown is available
  - Config kylin query pushdown to mysql, presto and hive according to https://github.com/apache/kylin/pull/1027.
  - Restart kylin and disable all the cube in `learn_kylin`
  - execute `select PART_DT, sum(price) from kylin_sales where PART_DT<'2012-01-15'` in project `learn_kylin`.
#### Expected result
- SQL is pushed down to presto and returns the correct result.

**Comment by zhangayqian** [comment]
### Verify test cases
- Case-1
  - Push down configuration
```
kylin.query.pushdown.runner-class-name=org.apache.kylin.query.adhoc.PushDownRunnerJdbcImpl
kylin.query.pushdown.update-enabled=false
kylin.query.pushdown.jdbc.url=jdbc:hive2://localhost:10000/default
kylin.query.pushdown.jdbc.driver=org.apache.hive.jdbc.HiveDriver
kylin.query.pushdown.jdbc.username=hive
kylin.query.pushdown.jdbc.password=
kylin.query.pushdown.jdbc.pool-max-total=8
kylin.query.pushdown.jdbc.pool-max-idle=8
kylin.query.pushdown.jdbc.pool-min-idle=0
```
  - Query log
image
  - Query Result
image
- Case-2
  - Push down configuration
```
kylin.query.pushdown.runner-class-name=org.apache.kylin.query.adhoc.PushDownRunnerJdbcImpl
kylin.query.pushdown.update-enabled=false
kylin.query.pushdown.jdbc.url=jdbc:presto://localhost:8083/hive/default
kylin.query.pushdown.jdbc.driver=com.facebook.presto.jdbc.prestodriver
kylin.query.pushdown.jdbc.username=presto
kylin.query.pushdown.jdbc.password=
kylin.query.pushdown.jdbc.pool-max-total=8
kylin.query.pushdown.jdbc.pool-max-idle=8
kylin.query.pushdown.jdbc.pool-min-idle=0
```
  - Query log
image
  - Query result
image

**Comment by zhangayqian** [comment]
### Verify test case-3 after bug fixed:
- Query Result
image
- Query log
image

**Comment by zhangayqian** [comment]
## Summary
- Single query engine(hive) pushdown is available.
- Single query engine(presto) pushdown is available.
- Multi Jdbc pushdown is available after bug fixed.

**Pull request merged**